src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -653,6 +653,99 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
+            foreach (var projectChange in projectChanges)
+            {
+                var documents = projectChange.GetChangedDocuments().Where(ShouldApplyChangesToMappedDocuments);
+                if (documents.Any())
+                {
+                    changedMappedDocuments.Add(projectChange, documents.ToImmutableArray());
+                }
+            }
+
+            if (changedMappedDocuments.Any())
+            {
+                ApplyChangesForMappedDocuments(changedMappedDocuments);
+            }
+
+            return;
+
+            bool ShouldApplyChangesToMappedDocuments(DocumentId id)
+            {
+                var document = this.CurrentSolution.GetDocument(id);
+                // Only consider files that are mapped and that we are unable to apply changes to.
+                // TODO - refactor how this is determined - https://github.com/dotnet/roslyn/issues/47908
+                return document?.Services?.GetService<ISpanMappingService>() != null && document?.CanApplyChange() == false;
+            }
+        }
+
+        private void ApplyChangesForMappedDocuments(Dictionary<ProjectChanges, ImmutableArray<DocumentId>> changedMappedDocuments)
+        {
+            // Get the original text changes from all documents and call the span mapping service to get span mappings for the text changes.
+            // Create mapped text changes using the mapped spans and original text changes' text.
+
+            // Mappings for opened razor files are retrieved via the LSP client making a request to the razor server.
+            // If we wait for the result on the UI thread, we will hit a bug in the LSP client that brings us to a code path
+            // using ConfigureAwait(true).  This deadlocks as it then attempts to return to the UI thread which is already blocked by us.
+            // Instead, we invoke this in JTF run which will mitigate deadlocks when the ConfigureAwait(true)
+            // tries to switch back to the main thread in the LSP client.
+            // Link to LSP client bug for ConfigureAwait(true) - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1216657
+            var mappedChanges = _threadingContext.JoinableTaskFactory.Run(() => GetMappedTextChanges(changedMappedDocuments));
+
+            // Group the mapped text changes by file, then apply all mapped text changes for the file.
+            foreach (var changesForFile in mappedChanges)
+            {
+                // It doesn't matter which of the file's projectIds we pass to the invisible editor, so just pick the first.
+                var projectId = changesForFile.Value.First().ProjectId;
+                // Make sure we only take distinct changes - we'll have duplicates from different projects for linked files or multi-targeted files.
+                var distinctTextChanges = changesForFile.Value.Select(change => change.TextChange).Distinct().ToImmutableArray();
```

Comment:
What happens if some of these edits are overlapping and producing different edits? I'm OK with us dropping on the floor some edits to attempt to make a best shot at something, or trying to incorporate some of the linked file logic we have, but this seems like this risks crashing.